pageserver/src/consumption_metrics.rs

```diff
@@ -471,6 +461,72 @@ where
     current_metrics
 }
 
+#[derive(thiserror::Error, Debug)]
+enum UploadError {
+    #[error("server rejected metrics")]
+    Rejected(reqwest::StatusCode),
+    #[error("request failed")]
+    Reqwest(#[source] reqwest::Error),
+    #[error("cancelled")]
+    Cancelled,
+}
+
+impl UploadError {
+    fn is_reject(&self) -> bool {
+        matches!(self, UploadError::Rejected(_))
+    }
+}
+
+async fn upload(
+    client: &reqwest::Client,
+    metric_collection_endpoint: &reqwest::Url,
+    body: bytes::Bytes,
+    cancel: &CancellationToken,
+) -> Result<(), UploadError> {
+    if cancel.is_cancelled() {
+        // retry does not drop the request when cancel comes, so pre-check it.
+        //
+        // it's not terrible, because those come at shutdown when we will have very little time to
+        // act anyways.
+        Err(UploadError::Cancelled)
+    } else {
+        utils::backoff::retry(
+            move || {
+                let body = body.clone();
+                async move {
+                    let res = client
+                        .post(metric_collection_endpoint.clone())
+                        .header(reqwest::header::CONTENT_TYPE, "application/json")
+                        .body(body)
+                        .send()
+                        .await;
+
+                    let res = res.and_then(|res| res.error_for_status());
+
+                    // TODO: ensure redirects are configured on
```

Comment:
```suggestion
                    // TODO: ensure redirection support is enabled
```

Reply:
It is on by default, I'll remove this comment later.

Reply:
Forgot with the last round, but I'll have conflicts for days with the enhancements already done.

Reply:
Ah, had to do another one for the changed strings. Included.